src/woo_publications/logging/logevent.py

```diff
@@ -0,0 +1,166 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from django.db import models
+    from ..accounts.models import User
+
+from .constants import Events
+
+__all__ = [
+    # admin
+    "audit_admin_create",
+    "audit_admin_read",
+    "audit_admin_update",
+    "audit_admin_delete",
+    # api
+    "audit_api_create",
+    "audit_api_read",
+    "audit_api_update",
+    "audit_api_delete",
+]
+
+"""must contain either a django_user or user_id + user_display"""
+
+
+def _audit_event(
+    content_object: models.Model,
+    event: Events,
+    user_id: str | None = None,
+    user_display: str | None = None,
+    django_user: User | None = None,
+    **kwargs: any,
+) -> None:
+
+    from .models import TimelineLogProxy
+
+    assert django_user or user_id and user_display
```

Comment:
Making this an explicit if/else branch or using typing.overload should allow you to narrow the types.

The type checker is correctly pointing out that django_user can still be None by the time you call get_full_name, it's unfortunately not able to narrow the type from this assert